### PR TITLE
add buildkite upload

### PR DESCRIPTION
**Pull request by finn-tbd**
and re-arranged some stuff:
* split out all report renders to their own file
* update result struct to track per-test runtime
* disable failing web5-kt test
* use web5-js test from web5-js repo, drop web5-js test from this repo
* renamed the go package to reflect the current repo name

### CODE REVIEW

File: .github/workflows/self-test.yaml [hunk 1]
```diff
@@ -10,11 +10,15 @@ jobs:
       matrix:
         SDK:
           - web5-js
-          - web5-kt
+          # - web5-kt
```
Comment (nitro-neal) [inline .github/workflows/self-test.yaml:13]: Did you mean to comment this out?

Comment (finn-tbd) [reply .github/workflows/self-test.yaml:13]: i did: it was failing for me, i assumed web5-kt had changed

Comment (finn-tbd) [reply .github/workflows/self-test.yaml:13]: looks like it failed in the latest commit on `main` as well

**Review by nitro-neal — APPROVED** [review]

**Review by andresuribe87 — COMMENTED** [review]
Should there be tests for the template producing pieces of the scaffolding?

File: test-harness/cmd/web5-test-harness/test-one.go [hunk 2]
```diff
@@ -129,3 +148,88 @@ func testOne(dir string) (reports.Report, error) {
...
+	}
+	defer resp.Body.Close()
+
+	responseBody := bytes.Buffer{}
+	if _, err := io.Copy(&responseBody, resp.Body); err != nil {
+		slog.Error("error reading response from buildkite")
+		return err
+	}
+
+	if resp.StatusCode > 299 {
```
[the remark this answers is not in the document]
Comment (finn-tbd) [reply test-harness/cmd/web5-test-harness/test-one.go:211]: I can't find any documentation on the expected error codes from buildkite for this endpoint, but > 299 will catch any normal error code. < 200 is "informational" ([according to MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#information_responses))

File: test-harness/cmd/web5-test-harness/test-one.go [hunk 3]
```diff
@@ -129,3 +148,88 @@ func testOne(dir string) (reports.Report, error) {
...
+
+	slog.Info("successfully uploaded results to buildkite", "response", body.String())
+
+	return nil
+}
+
+func addFormValue(writer *multipart.Writer, key, value string) error {
+	field, err := writer.CreateFormField(key)
+	if err != nil {
+		slog.Error("error creating form field", "key", key, "value", value)
```
[the remark this answers is not in the document]
Comment (finn-tbd) [reply test-harness/cmd/web5-test-harness/test-one.go:224]: I'm certainly familiar with that pattern, but personally prefer the the context messages related to the error to be logged on separate lines. I find it more readable. If the exact same message is getting repeated multiple times, that's an issue, but this is just adding context that may or may not be relevant to help understand where the error originated. I find the pattern of having an increasingly long error message with all the context on a single line to be confusing and hard to read.

If people have strong opinions, I'm happy to change this, but if it's my code and I'm going to have to debug it, I would prefer the approach i've adopted.

**Review by mistermoe — APPROVED** [review]

File: .github/workflows/pages.yaml [hunk 4]
```diff
@@ -11,7 +11,11 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v4
       - name: run for all SDKs
-        run: cd test-harness && go run ./cmd/web5-test-harness many sdks/* && mv _site ../_site
+        run: |
+          git clone https://github.com/TBD54566975/web5-js.git
```
Comment (decentralgabe) [inline .github/workflows/pages.yaml:15]: nit: let's favor ssh checkout over https

File: test-harness/cmd/web5-test-harness/test-one.go [hunk 5]
```diff
@@ -129,3 +148,88 @@ func testOne(dir string) (reports.Report, error) {
...
+	req.Header.Add("Authorization", fmt.Sprintf("Token token=%s", buildkiteToken))
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		slog.Error("error uploading results to buildkite")
+		return err
+	}
+	defer resp.Body.Close()
+
+	responseBody := bytes.Buffer{}
```
Comment (decentralgabe) [inline test-harness/cmd/web5-test-harness/test-one.go:205]:
```suggestion
	var responseBody bytes.Buffer
```

File: test-harness/cmd/web5-test-harness/test-one.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (decentralgabe) [reply test-harness/cmd/web5-test-harness/test-one.go:211]: we have a helper func in other libs that is like

```go
func is200Status(code int) bool {
  return code / 200 != 2 
}
```

File: test-harness/reports/html.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+type htmlTemplateInput struct {
+	Reports []Report
+	Tests   map[string][]string
+}
+
+func WriteHTML(reports []Report, filename string) error {
+	slog.Info("writing html report")
+
+	testmap := map[string]map[string]bool{}
```
Comment (decentralgabe) [inline test-harness/reports/html.go:27]: use `make` with maps

File: test-harness/reports/html.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+			for test := range tests {
+				testmap[category][test] = true
+			}
+		}
+	}
+
+	templateInput := htmlTemplateInput{Reports: reports, Tests: map[string][]string{}}
+
+	for category, tests := range testmap {
+		templateInput.Tests[category] = []string{}
```
Comment (decentralgabe) [inline test-harness/reports/html.go:43]: I believe with my suggestion above this is no longer necessary

**Comment by andresuribe87** [comment]
@finn-tbd just wanted to make sure you didn't miss https://github.com/TBD54566975/sdk-development/pull/68#pullrequestreview-1751550533

tl;dr; should all the code in here be tested? Apologies if it's being tested already and I missed it. 

Otherwise this lgtm

**Comment by finn-tbd** [comment]
@andresuribe87 this is all tested, most of it in the branch's CI:
* junit (new template) is rendered and sent to buildkite in CI because the token is present ([results here](https://buildkite.com/organizations/tbd-oss/analytics/suites/sdk-development?branch=all+branches))
* md is rendered and written to the github summary ([link to rendered md from the most recent pipeline](https://github.com/TBD54566975/sdk-development/actions/runs/7011455742?pr=68#summary-19074048900))
* text is rendered and displayed in the Github Actions text output (bottom of "self test" job from [this pipeline](https://github.com/TBD54566975/sdk-development/actions/runs/7011455742/job/19074048900?pr=68))
* html template i tested locally

File: .github/workflows/pages.yaml (hunk 4 above)
Comment (finn-tbd) [reply .github/workflows/pages.yaml:15]: this results in a failure ([ci output](https://github.com/TBD54566975/sdk-development/actions/runs/7022580060/job/19107178448?pr=68)):

image

**Comment by andresuribe87** [comment]
> edit: unless you're talking about more automated correctness verification like unit tests or something. I don't feel those are necessary right now, especially with the desire to increase our velocity, but if others disagree I'm open to looking at ways to unit test the templates.

I was talking about those automated tests. I'd expect things to be really easy to test if you already have samples. I would argue not having them decreases our velocity in the medium term, as modifications to anything in the upload pipeline will be hard to verify.
